### PR TITLE
Add ability to keep several IP addresses per PeerAddress

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
DNSSD discovery will discover potentially more than one IP address, yet existing code always chosses the first one to connect to.

#### Change overview
Makes PeerAddress able to keep more 'valid destinations' for sending packets. Intent is in the future to be able to try multiple addresses when sending: if one path times out, instead of giving up we will be able to move to the next possible address.

This change only adds ability to keep track of addresess and marks one instance of 'SetAddress' which seems wrong (does not set interface) to be replaced in the future.

#### Testing
Built and commissioned an all-cluster-app (used devkitc + linux chiptool).

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/controller/CommissioneeDeviceProxy.h [hunk 1]
```diff
@@ -216,7 +216,7 @@ class CommissioneeDeviceProxy : public DeviceProxy, public SessionReleaseDelegat
 
     Messaging::ExchangeManager * GetExchangeManager() const override { return mExchangeMgr; }
 
-    void SetAddress(const Inet::IPAddress & deviceAddr) { mDeviceAddress.SetIPAddress(deviceAddr); }
+    void SetAddress(const Inet::IPAddress & deviceAddr) { mDeviceAddress.SetSingleIPAddress(deviceAddr); }
```
Comment (andy31415) [inline src/controller/CommissioneeDeviceProxy.h:219]: This is the suspect case: SetAddress ignores what interface mDeviceAddress has.
So if this address is link-local, it is probably wrong, unless we have a SetInterface. We will need to trace what calls this and how and convert into some 'SetSingleDestination' or 'SetPeerAddress' instead.

Comment (tcarmelveilleux) [reply src/controller/CommissioneeDeviceProxy.h:219]: If you delete this method, what breaks? It 100% is a bad thing to ignore interface and we should not have any APIs that do so related to any discovery

File: src/transport/raw/PeerAddress.h [hunk 2]
```diff
@@ -121,79 +217,41 @@ class PeerAddress
...
 
-    static PeerAddress Uninitialized() { return PeerAddress(Inet::IPAddress::Any, Type::kUndefined); }
+    static PeerAddress Uninitialized() { return PeerAddress(Type::kUndefined); }
 
     static PeerAddress BLE() { return PeerAddress(Type::kBle); }
-    static PeerAddress UDP(const Inet::IPAddress & addr) { return PeerAddress(addr, Type::kUdp); }
-    static PeerAddress UDP(const Inet::IPAddress & addr, uint16_t port) { return UDP(addr).SetPort(port); }
-    static PeerAddress UDP(const Inet::IPAddress & addr, uint16_t port, Inet::InterfaceId interface)
+
+    static PeerAddress UDP(const Inet::IPAddress & addr, uint16_t port = CHIP_PORT,
```
Comment (tcarmelveilleux) [inline src/transport/raw/PeerAddress.h:241]: Please do not default to CHIP_PORT here

File: src/transport/raw/PeerAddress.h [hunk 3]
```diff
@@ -121,79 +217,41 @@ class PeerAddress
...
-        return UDP(addr).SetPort(port).SetInterface(interface);
+        PeerAddress result = PeerAddress(Type::kUdp);
+        result.SetPort(port).AppendDestination(addr, interface);
+        return result;
     }
-    static PeerAddress TCP(const Inet::IPAddress & addr) { return PeerAddress(addr, Type::kTcp); }
-    static PeerAddress TCP(const Inet::IPAddress & addr, uint16_t port) { return TCP(addr).SetPort(port); }
-    static PeerAddress TCP(const Inet::IPAddress & addr, uint16_t port, Inet::InterfaceId interface)
+
+    static PeerAddress TCP(const Inet::IPAddress & addr, uint16_t port = CHIP_PORT,
```
Comment (tcarmelveilleux) [inline src/transport/raw/PeerAddress.h:249]: Same comment about CHIP_PORT

File: src/transport/raw/PeerAddress.h [hunk 4]
```diff
@@ -209,14 +267,15 @@ class PeerAddress
...
 
 private:
-    Inet::IPAddress mIPAddress   = {};
-    Type mTransportType          = Type::kUndefined;
-    uint16_t mPort               = CHIP_PORT; ///< Relevant for UDP data sending.
-    Inet::InterfaceId mInterface = Inet::InterfaceId::Null();
+    IPDestination mDestinations[kMaxPeerDestinations] = {};
+    unsigned mNumValidDestinations                    = 0;
+    Type mTransportType                               = Type::kUndefined;
+    uint16_t mPort                                    = CHIP_PORT; ///< Relevant for UDP data sending.
```
Comment (tcarmelveilleux) [inline src/transport/raw/PeerAddress.h:278]: Need a TODO to not rely on CHIP_PORT get it from explicit setter and that it has to match the SRV record port

File: src/transport/raw/tests/TestPeerAddress.cpp [hunk 5]
```diff
@@ -62,18 +67,76 @@ void TestPeerAddressMulticast(nlTestSuite * inSuite, void * inContext)
...
+        IPAddress::FromString("1223::3456:789a", ip);
+        PeerAddress::UDP(ip, 8080).ToString(buff);
+
+        NL_TEST_ASSERT(inSuite, !strcmp(buff, "UDP:[1223::3456:789a]:8080"));
+    }
+
+    {
+
+        PeerAddress udp = PeerAddress(Transport::Type::kUdp);
+        udp.SetPort(5840);
```
Comment (tcarmelveilleux) [inline src/transport/raw/tests/TestPeerAddress.cpp:103]: Suggest a less confusing port like 23456. `5840` is close to default CHIP_PORT

File: src/controller/CommissioneeDeviceProxy.h (hunk 1 above)
Comment (andy31415) [reply src/controller/CommissioneeDeviceProxy.h:219]: I want to do that as a followup. I do not know what a 'Commissionee Device proxy' is ... I imagine I will have to track down the usage and at some point I should end up at a DNSSD lookup or some manual input somehow.

File: src/transport/raw/PeerAddress.h (hunk 2 above)
Comment (andy31415) [reply src/transport/raw/PeerAddress.h:241]: Existing code was already defaulting to this. I plan to do small changes at once.
The reason I claim  existing code did this: code before had a UDP(addr) implementation that was using CHIP_PORT.

File: src/transport/raw/PeerAddress.h (hunk 4 above)
Comment (andy31415) [reply src/transport/raw/PeerAddress.h:278]: This is existing code. I am looking for smaller deltas and don't want to go into "you touched it now you have to fix the whole thing". I am fixing things, but piecewise.

File: src/transport/raw/tests/TestPeerAddress.cpp (hunk 5 above)
Comment (andy31415) [reply src/transport/raw/tests/TestPeerAddress.cpp:103]: Does it matter?

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by andy31415** [comment]
This got merged, but I must say that BSS increase is concering. I believe just one IP address is insufficient, however I do plan to make it configurable and maybe allow small devices (which are often thread) to get by with just bare minimum like 2.

**Comment by msandstedt** [comment]
How is this actually going to help us?  Under what conditions would an address work, but then stop working?  And if this happens, don't we want to resolve again?

We do want to resolve all addresses, but #13274 not withstanding, the code does this.  The thing that is missing is logic to successively try addresses until one works.  Working would be defined as: can establish PASE to this address or can establish CASE to this address.

I actually have fully worked examples of this in #13294.

Anyway, back to this approach.  I think the only possible thing that this would do for is to allow for migration from ULAs after renumbering to GUAs.  That doesn't seem like great behavior.

**Comment by andy31415** [comment]
The intent is to allow trying more (ideally all but we seem to use too much RAM right now) of the IP addresses that are reported via DNSSD for nodes.

We seem to consistently see issues with routing and that some addresses work while some don't. Being able to do more than just 'pick first IPv6 address' should increase reliability until IPv6 works much better in consumer routers.

**Comment by msandstedt** [comment]
Yes, but we already have a structure that can hold multiple resolved addressed: ResolvedNodeData.  How does the fact that PeerId only holds one address prevent us from trying more than one?

**Comment by andy31415** [comment]
> Yes, but we already have a structure that can hold multiple resolved addressed: ResolvedNodeData. How does the fact that PeerId only holds one address prevent us from trying more than one?

@msandstedt 
We did not propagate them. There logic was to call a `ToPeerAddress` call that converts a ResolvedNodeData (multiple IPs) into a *single* IP address to use, so by the time CRMP figures out that the IP chosen does not seem to get replies, the ResolvedNodeData content is already lost.

Will try to reach out over slack, since Github-based chat probably does not work well for quick back and forth.

**Pull request merged**